In [3]:
import pandas as pd

In [9]:
df = pd.read_csv('C:\\Users\\bhatt\\OneDrive\\Desktop\\Bengaluru Rental Market Analysis\\data\\Bengaluru_rent.csv')

df.head()

,seller_type,bedroom,layout_type,property_type,locality,price,area,furnish_type,bathroom
0,OWNER,2,BHK,Independent Floor,kamakshipalya,"20,000",1140,Semi-Furnished,2 bathrooms
1,OWNER,2,BHK,Independent House,Kadugodi Industrial Area,"8,000",840,Semi-Furnished,2 bathrooms
2,OWNER,2,BHK,Independent House,Koramangala,"21,000",1000,Semi-Furnished,2 bathrooms
3,OWNER,1,BHK,Apartment,Sadduguntepalya,"9,000",550,Semi-Furnished,1 bathrooms
4,OWNER,3,BHK,Apartment,Chandapura,"17,000",1230,Furnished,3 bathrooms


In [10]:
df.shape

(23540, 9)

In [12]:
df.info()


<class 'pandas.DataFrame'>
RangeIndex: 23540 entries, 0 to 23539
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype
---  ------         --------------  -----
 0   seller_type    23540 non-null  str  
 1   bedroom        23540 non-null  int64
 2   layout_type    23540 non-null  str  
 3   property_type  23540 non-null  str  
 4   locality       23540 non-null  str  
 5   price          23540 non-null  str  
 6   area           23540 non-null  int64
 7   furnish_type   23540 non-null  str  
 8   bathroom       23106 non-null  str  
dtypes: int64(2), str(7)
memory usage: 1.6 MB


In [14]:
df.columns

Index(['seller_type', 'bedroom', 'layout_type', 'property_type', 'locality',
       'price', 'area', 'furnish_type', 'bathroom'],
      dtype='str')

In [15]:
df.isnull().sum()

seller_type        0
bedroom            0
layout_type        0
property_type      0
locality           0
price              0
area               0
furnish_type       0
bathroom         434
dtype: int64

In [19]:
df = df.dropna(subset=['bathroom'])
df.info()

<class 'pandas.DataFrame'>
Index: 23106 entries, 0 to 23539
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype
---  ------         --------------  -----
 0   seller_type    23106 non-null  str  
 1   bedroom        23106 non-null  int64
 2   layout_type    23106 non-null  str  
 3   property_type  23106 non-null  str  
 4   locality       23106 non-null  str  
 5   price          23106 non-null  str  
 6   area           23106 non-null  int64
 7   furnish_type   23106 non-null  str  
 8   bathroom       23106 non-null  str  
dtypes: int64(2), str(7)
memory usage: 1.8 MB


In [21]:
df.duplicated().sum()

np.int64(9136)

In [22]:
df[df.duplicated()].head(10)

,seller_type,bedroom,layout_type,property_type,locality,price,area,furnish_type,bathroom
100,OWNER,2,BHK,Independent House,Kadugodi Industrial Area,"8,000",840,Semi-Furnished,2 bathrooms
101,OWNER,2,BHK,Independent House,Kadugodi Industrial Area,"8,000",840,Semi-Furnished,2 bathrooms
102,OWNER,2,BHK,Independent House,Kadugodi Industrial Area,"8,000",840,Semi-Furnished,2 bathrooms
103,OWNER,2,BHK,Independent House,Kadugodi Industrial Area,"8,000",840,Semi-Furnished,2 bathrooms
104,OWNER,2,BHK,Independent House,Kadugodi Industrial Area,"8,000",840,Semi-Furnished,2 bathrooms
105,OWNER,2,BHK,Independent House,Koramangala,"21,000",1000,Semi-Furnished,2 bathrooms
106,OWNER,2,BHK,Independent House,Koramangala,"21,000",1000,Semi-Furnished,2 bathrooms
107,OWNER,2,BHK,Independent House,Koramangala,"21,000",1000,Semi-Furnished,2 bathrooms
108,OWNER,2,BHK,Independent House,Koramangala,"21,000",1000,Semi-Furnished,2 bathrooms
109,OWNER,2,BHK,Independent House,Koramangala,"21,000",1000,Semi-Furnished,2 bathrooms


Dropping duplicate values

In [24]:
df = df.drop_duplicates()

In [26]:
df.shape

(13970, 9)

In [28]:
df.duplicated().sum()

np.int64(0)

In [32]:
df.nunique()

seller_type         3
bedroom            10
layout_type         2
property_type       6
locality          876
price             594
area             1130
furnish_type        3
bathroom           20
dtype: int64

In [34]:
df.info()

<class 'pandas.DataFrame'>
Index: 13970 entries, 0 to 23525
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype
---  ------         --------------  -----
 0   seller_type    13970 non-null  str  
 1   bedroom        13970 non-null  int64
 2   layout_type    13970 non-null  str  
 3   property_type  13970 non-null  str  
 4   locality       13970 non-null  str  
 5   price          13970 non-null  str  
 6   area           13970 non-null  int64
 7   furnish_type   13970 non-null  str  
 8   bathroom       13970 non-null  str  
dtypes: int64(2), str(7)
memory usage: 1.1 MB


In [35]:
df

,seller_type,bedroom,layout_type,property_type,locality,price,area,furnish_type,bathroom
0,OWNER,2,BHK,Independent Floor,kamakshipalya,"20,000",1140,Semi-Furnished,2 bathrooms
1,OWNER,2,BHK,Independent House,Kadugodi Industrial Area,"8,000",840,Semi-Furnished,2 bathrooms
2,OWNER,2,BHK,Independent House,Koramangala,"21,000",1000,Semi-Furnished,2 bathrooms
3,OWNER,1,BHK,Apartment,Sadduguntepalya,"9,000",550,Semi-Furnished,1 bathrooms
4,OWNER,3,BHK,Apartment,Chandapura,"17,000",1230,Furnished,3 bathrooms
...,...,...,...,...,...,...,...,...,...
23521,OWNER,2,BHK,Apartment,Electronic City Phase 2,"14,000",850,Semi-Furnished,2 bathrooms
23522,OWNER,2,BHK,Apartment,Bellandur,"18,000",1160,Semi-Furnished,2 bathrooms
23523,OWNER,2,BHK,Apartment,Horamavu,"13,000",1070,Semi-Furnished,2 bathrooms
23524,OWNER,1,BHK,Independent House,Bommanahalli,"7,500",450,Semi-Furnished,1 bathrooms


Changing the data types for prices from str to int as it is an aggregation data metric.

In [ ]:
df['price'] = df['price'].str.replace('₹', '').str.replace(',', '').astype(float)

In [42]:
df

,seller_type,bedroom,layout_type,property_type,locality,price,area,furnish_type,bathroom
0,OWNER,2,BHK,Independent Floor,kamakshipalya,20000.0,1140,Semi-Furnished,2 bathrooms
1,OWNER,2,BHK,Independent House,Kadugodi Industrial Area,8000.0,840,Semi-Furnished,2 bathrooms
2,OWNER,2,BHK,Independent House,Koramangala,21000.0,1000,Semi-Furnished,2 bathrooms
3,OWNER,1,BHK,Apartment,Sadduguntepalya,9000.0,550,Semi-Furnished,1 bathrooms
4,OWNER,3,BHK,Apartment,Chandapura,17000.0,1230,Furnished,3 bathrooms
...,...,...,...,...,...,...,...,...,...
23521,OWNER,2,BHK,Apartment,Electronic City Phase 2,14000.0,850,Semi-Furnished,2 bathrooms
23522,OWNER,2,BHK,Apartment,Bellandur,18000.0,1160,Semi-Furnished,2 bathrooms
23523,OWNER,2,BHK,Apartment,Horamavu,13000.0,1070,Semi-Furnished,2 bathrooms
23524,OWNER,1,BHK,Independent House,Bommanahalli,7500.0,450,Semi-Furnished,1 bathrooms


In [45]:
df['bedroom'] = df['bedroom'].astype(int)

In [46]:
df.info()

<class 'pandas.DataFrame'>
Index: 13970 entries, 0 to 23525
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   seller_type    13970 non-null  str    
 1   bedroom        13970 non-null  int64  
 2   layout_type    13970 non-null  str    
 3   property_type  13970 non-null  str    
 4   locality       13970 non-null  str    
 5   price          13970 non-null  float64
 6   area           13970 non-null  int64  
 7   furnish_type   13970 non-null  str    
 8   bathroom       13970 non-null  str    
dtypes: float64(1), int64(2), str(6)
memory usage: 1.1 MB


In [ ]:
df['bathroom'] = df['bathroom'].str.extract('(\d+)')

<>:1: SyntaxWarning: "\d" is an invalid escape sequence. Such sequences will not work in the future. Did you mean "\\d"? A raw string is also an option.
<>:1: SyntaxWarning: "\d" is an invalid escape sequence. Such sequences will not work in the future. Did you mean "\\d"? A raw string is also an option.
C:\Users\bhatt\AppData\Local\Temp\ipykernel_27380\2630738444.py:1: SyntaxWarning: "\d" is an invalid escape sequence. Such sequences will not work in the future. Did you mean "\\d"? A raw string is also an option.
  df['bathroom'] = df['bathroom'].str.extract('(\d+)')


,seller_type,bedroom,layout_type,property_type,locality,price,area,furnish_type,bathroom
0,OWNER,2,BHK,Independent Floor,kamakshipalya,20000.0,1140,Semi-Furnished,2
1,OWNER,2,BHK,Independent House,Kadugodi Industrial Area,8000.0,840,Semi-Furnished,2
2,OWNER,2,BHK,Independent House,Koramangala,21000.0,1000,Semi-Furnished,2
3,OWNER,1,BHK,Apartment,Sadduguntepalya,9000.0,550,Semi-Furnished,1
4,OWNER,3,BHK,Apartment,Chandapura,17000.0,1230,Furnished,3
...,...,...,...,...,...,...,...,...,...
23521,OWNER,2,BHK,Apartment,Electronic City Phase 2,14000.0,850,Semi-Furnished,2
23522,OWNER,2,BHK,Apartment,Bellandur,18000.0,1160,Semi-Furnished,2
23523,OWNER,2,BHK,Apartment,Horamavu,13000.0,1070,Semi-Furnished,2
23524,OWNER,1,BHK,Independent House,Bommanahalli,7500.0,450,Semi-Furnished,1


In [ ]:
df.info()

<class 'pandas.DataFrame'>
Index: 13970 entries, 0 to 23525
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   seller_type    13970 non-null  str    
 1   bedroom        13970 non-null  int64  
 2   layout_type    13970 non-null  str    
 3   property_type  13970 non-null  str    
 4   locality       13970 non-null  str    
 5   price          13970 non-null  float64
 6   area           13970 non-null  int64  
 7   furnish_type   13970 non-null  str    
 8   bathroom       13949 non-null  str    
dtypes: float64(1), int64(2), str(6)
memory usage: 1.1 MB


In [ ]:
df.isnull().sum()

seller_type       0
bedroom           0
layout_type       0
property_type     0
locality          0
price             0
area              0
furnish_type      0
bathroom         21
dtype: int64

Now remove the null values if the number accounts for less than 5% of the total dataset.

In [52]:
df.dropna(subset=['bathroom'], inplace=True)

In [54]:
df['bathroom']= df['bathroom'].astype(int)

df.info()

<class 'pandas.DataFrame'>
Index: 13949 entries, 0 to 23525
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   seller_type    13949 non-null  str    
 1   bedroom        13949 non-null  int64  
 2   layout_type    13949 non-null  str    
 3   property_type  13949 non-null  str    
 4   locality       13949 non-null  str    
 5   price          13949 non-null  float64
 6   area           13949 non-null  int64  
 7   furnish_type   13949 non-null  str    
 8   bathroom       13949 non-null  int64  
dtypes: float64(1), int64(3), str(5)
memory usage: 1.1 MB


No we calculate the numerical representation for each int type to understand the distribution.

In [55]:
df.describe()

,bedroom,price,area,bathroom
count,13949.000000,13949.000000,13949.000000,13949.000000
mean,1.980859,19939.756730,1161.799054,1.882357
std,0.857493,13249.159661,713.922610,0.883927
min,1.000000,1.000000,100.000000,1.000000
25%,1.000000,11000.000000,700.000000,1.000000
50%,2.000000,16500.000000,1100.000000,2.000000
75%,2.000000,25000.000000,1370.000000,2.000000
max,15.000000,99000.000000,10800.000000,19.000000


The distribution shows: 
# minimum price to be 1 rs
# max rooms as 15
# area at 10800 sq ft
# bathrooms at 19

We have to analyze each anomaly that seems like and outlier.

# i.e Rent could never be 1 rs which could be an outlier requiring further analysis.

We have to look for values which have the price at 1 and remove if they seem suspicious or out of place, similarly for other values we have to look for a reasonable explanation.

# i.e A 13 bedroom Villa could have 19 bathroom with 10800 sq ft area but the bathrom count of 19 is also visible for a 2bhk apartment then it's an outlier.

In [77]:
df.sort_values(by='price').head(200)

,seller_type,bedroom,layout_type,property_type,locality,price,area,furnish_type,bathroom
17433,AGENT,3,BHK,Villa,Bellandur,1.0,2350,Semi-Furnished,4
13504,AGENT,4,BHK,Apartment,RT Nagar Main Road,1.0,4000,Furnished,4
17420,AGENT,3,BHK,Apartment,Hebbal,1.0,2526,Furnished,4
3404,AGENT,3,BHK,Independent House,HSR Layout,1.0,3500,Semi-Furnished,3
21775,OWNER,2,BHK,Independent House,Vimanapura,1.0,700,Semi-Furnished,1
...,...,...,...,...,...,...,...,...,...
12115,AGENT,4,BHK,Apartment,Ashok Nagar,8.0,7000,Semi-Furnished,5
13254,OWNER,10,BHK,Independent Floor,Koramangala,9.0,9500,Furnished,4
18599,OWNER,1,BHK,Independent Floor,Nagavara,1500.0,250,Unfurnished,1
59,OWNER,1,RK,Studio Apartment,J. P. Nagar,1700.0,250,Unfurnished,1


In [78]:
print((df['price'] <= 10).sum())
print(((df['price'] > 10) & (df['price'] < 1000)).sum())

197
0


In [ ]:
df[df['price'] <= 10]['price'].value_counts().sort_index()

price
1.00    31
1.05     2
1.09     1
1.10    18
1.12     1
1.15     2
1.19     4
1.20    13
1.25    14
1.28     1
1.30     6
1.40     7
1.45     2
1.50    21
1.55     1
1.60     5
1.62     1
1.65     2
1.70     5
1.75     2
1.80     4
1.85     1
1.90     3
2.00    11
2.05     1
2.25     1
2.50    11
2.70     1
2.75     1
2.94     1
3.00    13
3.30     1
3.75     1
4.12     1
4.50     1
5.00     3
6.00     1
8.00     1
9.00     1
Name: count, dtype: int64

Here the data distribution requires more investigation as it is a possibility that the values 1 to 10 might represent a different number.

In [84]:
df[(df['price'] <= 10) & (df['bedroom'] <= 3)][['price','bedroom','area','locality']].sort_values('price')

,price,bedroom,area,locality
2530,1.00,3,3200,Varthur
2342,1.00,3,3600,Varthur
9101,1.00,3,2400,Frazer Town
3404,1.00,3,3500,HSR Layout
12063,1.00,3,2050,Lavelle Road
12402,1.00,3,3200,Koramangala
13714,1.00,3,2372,Frazer Town
10151,1.00,3,2800,Frazer Town
12458,1.00,3,2797,Frazer Town
17433,1.00,3,2350,Bellandur


During the analysis for these 197 outliers it was observed that most of these values had been representing different meaning for prices lower than 10.

Even if these prices were somehow reported less by a calculation error (i.e 10000x or 100000x part of the actual rent), these properties do not hold information that could be relied upon due to the imbalanced rates for areas like Varthur or Vimanapura as these areas represnt a complete different range for renting properties.

It is safe to delete these values as an error.

In [86]:
df = df[df['price'] > 10]

In [90]:
df[df['price'] > 10]['price'].min()

np.float64(1500.0)

As we have removed all the values for the outlier we can finally see a dataset which is representing accurate numbers.

We need to add the rent per sq feet coluimn for all the properties that will provide a deeper insight into locality wise prperty values that can help a person identify the best property for them to choose.   

In [91]:
df['rent_per_sq_ft'] = df['price'] / df['area']

df.head()

,seller_type,bedroom,layout_type,property_type,locality,price,area,furnish_type,bathroom,rent_per_sq_ft
0,OWNER,2,BHK,Independent Floor,kamakshipalya,20000.0,1140,Semi-Furnished,2,17.543860
1,OWNER,2,BHK,Independent House,Kadugodi Industrial Area,8000.0,840,Semi-Furnished,2,9.523810
2,OWNER,2,BHK,Independent House,Koramangala,21000.0,1000,Semi-Furnished,2,21.000000
3,OWNER,1,BHK,Apartment,Sadduguntepalya,9000.0,550,Semi-Furnished,1,16.363636
4,OWNER,3,BHK,Apartment,Chandapura,17000.0,1230,Furnished,3,13.821138


In [92]:
df.rent_per_sq_ft.describe()

count    13752.000000
mean        18.631064
std          7.503579
min          4.000000
25%         13.941579
50%         17.582875
75%         22.000000
max        175.000000
Name: rent_per_sq_ft, dtype: float64

In [93]:
df.nlargest(10, 'rent_per_sq_ft')[['price', 'area', 'rent_per_sq_ft', 'locality']]

,price,area,rent_per_sq_ft,locality
13764,17500.0,100,175.000000,Shantipura Main
22965,30000.0,200,150.000000,Bellandur
23230,25500.0,250,102.000000,Magrath Road
20524,50000.0,500,100.000000,Jayanagar
2211,18000.0,200,90.000000,Seshadripuram
19598,50500.0,600,84.166667,Ashok Nagar
21451,35000.0,420,83.333333,Electronics City
21088,18000.0,220,81.818182,HSR Layout
17891,12000.0,150,80.000000,Nagarbhavi
19196,12000.0,150,80.000000,Murugeshpalya
